kernel/src/transaction.rs

```diff
@@ -112,6 +138,107 @@ impl Transaction {
         self.commit_info = Some(commit_info.into());
         self
     }
+
+    // Generate the logical-to-physical transform expression which must be evaluated on every data
+    // chunk before writing. At the moment, this is a transaction-wide expression.
+    fn generate_logical_to_physical(&self) -> Expression {
+        // for now, we just pass through all the columns except partition columns.
+        // note this is _incorrect_ if table config deems we need partition columns.
+        Expression::struct_from(self.read_snapshot.schema().fields().filter_map(|f| {
+            if self
+                .read_snapshot
+                .metadata()
+                .partition_columns
+                .contains(f.name())
+            {
+                None
+            } else {
+                let col_name = ColumnName::new([f.name()]);
+                Some(col_name.into())
+            }
+        }))
+    }
+
+    /// Get the write context for this transaction. At the moment, this is constant for the whole
+    /// transaction.
+    // Note: after we introduce metadata updates (modify table schema, etc.), we need to make sure
+    // that engines cannot call this method after a metadata change, since the write context could
+    // have invalid metadata.
+    pub fn get_write_context(&self) -> WriteContext {
+        let target_dir = self.read_snapshot.table_root();
+        let snapshot_schema = self.read_snapshot.schema();
+        let logical_to_physical = self.generate_logical_to_physical();
+        WriteContext::new(
+            target_dir.clone(),
+            Arc::new(snapshot_schema.clone()),
+            logical_to_physical,
+        )
+    }
+
+    /// Add write metadata about files to include in the transaction. This API can be called
+    /// multiple times to add multiple batches.
+    ///
+    /// The expected schema for `write_metadata` is given by [`get_write_metadata_schema`].
+    pub fn add_write_metadata(&mut self, write_metadata: Box<dyn EngineData>) {
+        self.write_metadata.push(write_metadata);
+    }
+}
+
+// convert write_metadata into add actions using an expression to transform the data in a single
+// pass
+fn generate_adds<'a>(
+    engine: &dyn Engine,
+    write_metadata: impl Iterator<Item = &'a dyn EngineData> + Send + 'a,
+) -> impl Iterator<Item = DeltaResult<Box<dyn EngineData>>> + Send + 'a {
+    let expression_handler = engine.get_expression_handler();
+    let write_metadata_schema = get_write_metadata_schema();
+    let log_schema = get_log_add_schema();
+
+    write_metadata.map(move |write_metadata_batch| {
+        let adds_expr = Expression::struct_from([Expression::struct_from(
+            write_metadata_schema
+                .fields()
+                .map(|f| ColumnName::new([f.name()]).into()),
+        )]);
+        let adds_evaluator = expression_handler.get_evaluator(
+            write_metadata_schema.clone(),
+            adds_expr,
+            log_schema.clone().into(),
```

Comment:
`into` what? Isn't it already a `SchemaRef`?

Reply:
`Arc<StructType>` into our `Datatype`